prune/prune.go

```diff
@@ -16,184 +16,249 @@ package prune
 
 import (
 	"context"
+	"errors"
 	"fmt"
-	"time"
 
 	"github.com/go-logr/logr"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/runtime/schema"
-	"k8s.io/client-go/kubernetes"
 
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
-)
 
-// ResourceStatus describes the Kubernetes resource status we are evaluating
-type ResourceStatus string
-
-// Strategy describes the pruning strategy we want to employ
-type Strategy string
-
-const (
-	// CustomStrategy maximum age of a resource that is desired, Duration
-	CustomStrategy Strategy = "Custom"
-	// MaxAgeStrategy maximum age of a resource that is desired, Duration
-	MaxAgeStrategy Strategy = "MaxAge"
-	// MaxCountStrategy maximum number of a resource that is desired, int
-	MaxCountStrategy Strategy = "MaxCount"
-	// JobKind equates to a Kube Job resource kind
-	JobKind string = "Job"
-	// PodKind equates to a Kube Pod resource kind
-	PodKind string = "Pod"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
 )
 
-// StrategyConfig holds settings unique to each pruning mode
-type StrategyConfig struct {
-	Mode            Strategy
-	MaxAgeSetting   string
-	MaxCountSetting int
-	CustomSettings  map[string]interface{}
+/*
+	-------------------------------------------
+	New Auto Pruning API Implementation
+	-------------------------------------------
+*/
+
+// Pruner is an object that runs a prune job.
+type Pruner struct {
+	Registry
+
+	// Client is the k8s client that will be used
+	Client client.Client
+
+	// DryRun indicates whether or not we should actually perform pruning or just return the list of pruneable objects
+	// true = Just check, don't prune
+	// false (default) = Prune
+	DryRun bool
```

Comment:
I didn't notice that there was a `DryRunClient`, and since we are already having them pass in a client it would probably be better than setting a `DryRun` value on the `Pruner` itself. I think as an added bonus it makes the internal pruner logic more straightforward as we don't have to worry about that `DryRun` parameter. I will go ahead and make the corresponding changes.

Reply:
Done in https://github.com/operator-framework/operator-lib/pull/105/commits/3f60543cf9ce9403d249d252ce3cf1bd9a42d34d